### ISSUE TITLE
Extrinsic ID (hash) of currently executing extrinsic should be accesible in the runtime

### ISSUE THREAD

**Issue by Sushisource** [paritytech/substrate#5153]
Someone submitting an extrinsic can easily obtain the id/hash of that extrinsic before submitting it to a validator by hashing it. (Side note: Why isn't this exposed as `.hash()` on [UncheckedExtrinsic](https://substrate.dev/rustdocs/pre-v2.0-3e65111/sp_runtime/generic/struct.UncheckedExtrinsic.html) while it is on `Block` and `Header`?)

However, from within the runtime, one cannot easily determine the hash of the extrinsic currently being processed. This is important if, for example, some external system refers to extrinsics by these IDs. In our case, some extrinsics create a pending state, which is later resolved by another extrinsic. We could create our own identifiers for this purpose (and is what we do currently), but given extrinsics already have a unique ID it would be desirable to re-use this. However, I can't create a map of <extrinsic id, data> in state without being able to access the ID, hence the need.

If there's some workaround for getting this now, I would love to know. Alternatively, if it can't be done, I'd be happy to look into making a PR for the functionality.

**Comment by tomusdrw** [comment]
Extrinsics can be encoded and hashed within the runtime if needed, we are encoding them already here:
https://github.com/paritytech/substrate/blob/3ca7ccef985407105d525b3ca2d1a79a820f3318/frame/executive/src/lib.rs#L280

Indeed from a pallet perspective, there is no way to tell - on that level we are operating on `Call` already rather than transactions.

Note also that hashes of extrinsics are not necessarily unique - there are no assumptions in the entire client code that the same extrinsic cannot be included twice (it's up to the runtime to decide). Even with FRAME when the account drops below existential deposit some extrinsics can be re-executed after the account is topped up (since on account reaping the nonce will be reset). So overall I'd be careful with relying on generic extrinsic hash as a unique identifier.
That said we use that in transaction pool, as at the same time there can only be one such transaction.

For in-block extrinsics I think `(blockhash, extrinsic index)` is a better identifier or addressing scheme.

**Comment by Sushisource** [comment]
@tomusdrw In this case we would explicitly be preventing duplicates. And, unfortunately, the other suggested identifier cannot work because we have the need for the user submitting the transaction to be able to know its id before it is submitted (hence neither blockhash or index may be known).

If I made a PR that piped the encoded extrinsic such that it was available within a pallet, would that be a change that would be welcome?

**Comment by gavofyork** [comment]
> @tomusdrw In this case we would explicitly be preventing duplicates. And, unfortunately, the other suggested identifier cannot work because we have the need for the user submitting the transaction to be able to know its id before it is submitted (hence neither blockhash or index may be known).
> 
> If I made a PR that piped the encoded extrinsic such that it was available within a pallet, would that be a change that would be welcome?
> 
> Tangential question: Doesn't the account index and the pubkey being part of the extrinsic guarantee uniqueness? I'm just curious about the possibility of non-uniqueness among extrinsic IDs you mentioned, seems like something that probably should be mentioned in https://substrate.dev/docs/en/conceptual/node/extrinsics

No. As @tomusdrw said, accounts whose balances drop too low can be reaped, reseting the account index.

And no the change isn't especially welcome. Extrinsics can be huge (some are 1.5MB +). Casually copying them around into new bits of storage is not a change to be made lightly given the resource-constrained environment. This is especially true if you (a) only case about getting the hash anyway (it could be determined and passed in prior to decode, or, at least, calculated while the extrinsic is still in scope, reducing the need to copy anything into temporary storage); and (b) it's not at all clear that the hash would be useful for your needs given that it is not guaranteed to be unique.

**Comment by Sushisource** [comment]
@gavofyork Thanks for the info. I understand the non-uniqueness claim now (although we have reaping handled differently in our system so I'm not sure it applies, but I'd still rather follow best practices). 

A bigger question I'm curious about here is what substrate implementers are expected to use for a unique transaction identifier. Is the expectation simply that, if you have a need for one, you should create your own scheme? Is there a more canonical way to obtain a unique transaction identifier? It's not really that I'm looking to identify a specific operation, I'm just looking to identify the initiating extrinsic/transaction.

This rings as pretty important to me to call out in the docs. We proceeded under the assumption that the extrinsic hash was unique as this is lightly implied in a handful of places. For example in this rpc api [here](https://polkadot.js.org/api/substrate/rpc.html#removeextrinsic-bytesorhash-vec-extrinsicorhash-vec-hash).

Thanks again!
